src/sync.rs

```diff
@@ -554,6 +588,117 @@ where
     }
 }
 
+/// Executes a closure with a Python critical section held on a `PyMutex`.
+///
+/// Acquires the mutex `mutex` until the closure `f` finishes. The mutex may be
+/// temporarily released and re-acquired if the closure calls back into the
+/// interpreter in a manner that would block.  This is similar to how the GIL
+/// can be released during blocking calls. See the safety notes below for
+/// caveats about releasing critical sections.
+///
+/// This is structurally equivalent to the use of the paired
+/// Py_BEGIN_CRITICAL_SECTION_MUTEX and Py_END_CRITICAL_SECTION C-API macros.
+///
+/// A no-op on GIL-enabled builds, where the critical section API is exposed as
+/// a no-op by the Python C API.
+///
+/// This variant is particularly useful when paired with a global `PyMutex` to
+/// create a "local GIL" to protect global state in an extension in an
+/// analogous manner to the GIL without introducing any deadlock risks or
+/// affecting runtime behavior on the GIL-enabled build.
+///
+/// # Safety
+///
+/// Provides weaker locking guarantees than traditional locks, but can in some
+/// cases be used to provide guarantees similar to the GIL without the risk of
+/// deadlocks associated with traditional locks.
+///
+/// The caller must ensure the closure cannot implicitly release the critical
+/// section. If a multithreaded program calls back into the Python interpreter
+/// in a manner that would cause the critical section to be released, the
+/// `PyMutex` will be released and the resource protected by the `PyMutex` may
+/// be read from or modified by another thread. Concurrent modifications are
+/// still impossible, but the state of the resource may change "underneath" a
+/// suspended thread in possibly surprising ways.
+///
+/// Only available on Python 3.14 and newer.
+#[cfg(all(Py_3_14, not(Py_LIMITED_API)))]
+#[cfg_attr(not(Py_GIL_DISABLED), allow(unused_variables))]
+pub fn with_critical_section_mutex<'py, F, R, T>(_py: Python<'py>, mutex: &PyMutex<T>, f: F) -> R
+where
+    F: FnOnce(&UnsafeCell<T>) -> R,
```

Comment:
One way to make the "safety" of this clearer would be to package the `UnsafeCell` up as something like

```rust
struct EnteredCriticalSection<'a, T>(&'a UnsafeCell<T>);

impl<T> EnteredCriticalSection<'_, T> {
    unsafe fn get(&mut self) -> &mut T { ... }
}
```

Can then document safety invariants on the `get` function. It's also a bit nicer than users having to go through `UnsafeCell::get` themselves (and a raw pointer).